### PR TITLE
Cant fetch contract info fro GCed heights

### LINKED ISSUE #3569: Can't fetch contract call info
This happens on GC enabled nodes at height that had already been GCed. The source of the problem is in [expectation](https://github.com/aeternity/aeternity/blob/master/apps/aecore/src/aec_trees.erl#L427) that the account's root hash at that height is present.

Stacktrace:
`Ranch listener external, connection process <0.21653.236>, stream 1 had its request process <0.21451.236> exit with reason {hash_not_present_in_db,<<139,205,230,170,134,106,66,142,33,150,217,218,72,88,33,76,31,60,242,213,255,185,15,254,173,65,47,249,225,122,166,60>>} and stacktrace [{aeu_mp_trees,db_get,2,[{file,"/home/builder/aeternity/apps/aeutils/src/aeu_mp_trees.erl"},{line,1110}]},{aeu_mp_trees,new,2,[{file,"/home/builder/aeternity/apps/aeutils/src/aeu_mp_trees.erl"},{line,128}]},{aec_trees,deserialize_from_db,2,[{file,"/home/builder/aeternity/apps/aecore/src/aec_trees.erl"},{line,427}]},{aec_db,find_block_state,2,[{file,"/home/builder/aeternity/apps/aecore/src/aec_db.erl"},{line,684}]},{aec_chain,get_block_state,1,[{file,"/home/builder/aeternity/apps/aecore/src/aec_chain.erl"},{line,562}]},{aec_chain,get_contract_call,3,[{file,"/home/builder/aeternity/apps/aecore/src/aec_chain.erl"},{line,270}]},{aehttp_helpers,'-get_info_object_from_tx/3-fun-0-',5,[{file,"/home/builder/aeternity/apps/aehttp/src/aehttp_helpers.erl"},{line,288}]},{aehttp_helpers,process_request,3,[{file,"/home/builder/aeternity/apps/aehttp/src/aehttp_helpers.erl"},{line,52}]}]
`

### PR DESCRIPTION
**Pull request by velzevur**
Fixes #3569

Allows loading only needed tries for lookups.

### CODE REVIEW

File: apps/aecore/src/aec_trees.erl [hunk 1]
```diff
@@ -417,24 +427,60 @@ deserialize_from_db(Bin, DirtyBackend) when is_binary(Bin), is_boolean(DirtyBack
...
+    lists:foldl(
+        fun(Element, AccumTrees) ->
+            case Element of
+                contracts ->
+                    ContractsT =
+                        case DirtyBackend of
+                            false -> aect_state_tree:new_with_backend(Contracts);
+                            true -> aect_state_tree:new_with_dirty_backend(Contracts)
+                        end,
+                    AccumTrees#trees{contracts = ContractsT};
```
Comment (hanssv) [inline apps/aecore/src/aec_trees.erl:439]: Maybe something like:
```
Backend = case DirtyBackend of false -> new_with_backend; true -> new_with_dirty_backend end,
...
AccTrees#trees{contracts = aect_state_tree:Backend(Contracts)};
```
Should be shorter, and perhaps more readable? 🤔

**Review by hanssv — APPROVED** [review]
LGTM, some improvement suggestions...

File: apps/aecore/src/aec_chain.erl [hunk 2]
```diff
@@ -103,7 +103,7 @@ get_account(PubKey) ->
     end.
 
 get_account_at_hash(PubKey, Hash) ->
-    case get_block_state(Hash) of
+    case get_block_state_partial(Hash, [accounts]) of
```
Comment (gorbak25) [inline apps/aecore/src/aec_chain.erl:106]: Switch the trees to a dirty backend :)
Here when we walk the MPT we generally access every MPT node as a full DB transaction.

File: apps/aecore/src/aec_trees.erl [hunk 3]
```diff
@@ -623,9 +669,9 @@ apply_txs_on_state_trees([SignedTx | Rest], ValidTxs, InvalidTxs, Trees, Env, Op
                     Reason = {Type, What},
                     lager:error("Tx ~p cannot be applied due to an error ~p", [Tx, Reason]),
                     {error, Reason};
-                Type:What:_ST when not Strict ->
+                Type:What:ST when not Strict ->
                     Reason = {Type, What},
-                    %%lager:debug("Stacktrace: ~p", [ST]),
+                    lager:debug("Stacktrace: ~p", [ST]),
```
Comment (gorbak25) [inline apps/aecore/src/aec_trees.erl:674]: Don't request the stack trace in the hot path :)

**Review by hanssv — APPROVED** [review]

**Review by cytadela8 — APPROVED** [review]

**Pull request merged**